GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -2145,7 +2145,512 @@ subroutine read_obs_sm_ASCAT_EUMET(                                         &
+
+    ! construct the file name using hardwired name and date_time
+
+    tmpname = 'cyg.ddmi.s'// YYYY // MM // DD //'-030000-e'// YYYY // MM // DD //'-210000.l3.grid-soil-moisture-36km.a32.d33.nc'
+        
+    tmpfname = trim(this_obs_param%path) // '/Y' // YYYY // '/M' // MM // '/' // trim(tmpname)
+
+    if (logit) write (logunit, '(400A)') 'Reading CYGNSS soil moisture data from file: ', trim(tmpfname)
+
+    ! check if file exists
+
+    inquire(file=tmpfname, exist=file_exists)
+
+    if (.not. file_exists) then
+       err_msg = 'CYGNSS SM obs file not found!'
+       call ldas_abort(LDAS_GENERIC_ERROR, Iam, err_msg)
+    end if
+
+    ! open the NetCDF observation file
+    ierr = nf90_open(trim(tmpfname), nf90_nowrite, ncid)
+
+    ! get variable dimension IDs
+    ierr = nf90_inq_dimid(ncid, 'lon',        lon_dimid)
+    ierr = nf90_inq_dimid(ncid, 'lat',        lat_dimid)
+    ierr = nf90_inq_dimid(ncid, 'time',       time_dimid)    
+    ierr = nf90_inq_dimid(ncid, 'timeslices', timeslices_dimid)
+    ierr = nf90_inq_dimid(ncid, 'startstop',  startstop_dimid)
+
+    ! dimensions sizes
+    ierr = nf90_inquire_dimension(ncid, lon_dimid,        len=N_lon)
+    ierr = nf90_inquire_dimension(ncid, lat_dimid,        len=N_lat)
+    ierr = nf90_inquire_dimension(ncid, time_dimid,       len=N_time)
+    ierr = nf90_inquire_dimension(ncid, timeslices_dimid, len=N_timeslices)
+    ierr = nf90_inquire_dimension(ncid, startstop_dimid,  len=N_startstop)
+
+    ! get variable IDs
+    ierr = nf90_inq_varid(ncid, 'SM_daily',       sm_d_varid)
+    ierr = nf90_inq_varid(ncid, 'SM_subdaily',    sm_subd_varid)
+    ierr = nf90_inq_varid(ncid, 'SIGMA_daily',    sigma_d_varid)
+    ierr = nf90_inq_varid(ncid, 'SIGMA_subdaily', sigma_subd_varid)
+    ierr = nf90_inq_varid(ncid, 'timeintervals',  timeintervals_varid)
+    ierr = nf90_inq_varid(ncid, 'latitude',       lat_varid)
+    ierr = nf90_inq_varid(ncid, 'longitude',      lon_varid)
+
+    ! allocate memory for universal variables
+    allocate(timeintervals(N_startstop,  N_timeslices ))
+    allocate(latitudes(    N_lon, N_lat               ))
+    allocate(longitudes(   N_lon, N_lat               ))
+    allocate(tmp_sm(       N_lon, N_lat               ))
+    allocate(tmp_sigma(    N_lon, N_lat               ))
+
+    ! define the count array for the NetCDF read operations
+    count = (/ N_lon, N_lat, 1 /)
+    
+    ! read the variables
+    ierr = nf90_get_var(ncid, timeintervals_varid, timeintervals)
+    ierr = nf90_get_var(ncid, lat_varid, latitudes)
+    ierr = nf90_get_var(ncid, lon_varid, longitudes)
+
+    idx = -1
+
+    ! read either subdaily or daily soil moisture and sigma variables
+    
+    if (dt_obs == 6) then
+       do i = 1, N_timeslices
+          if ((obs_hour ==  3 .and. (abs(timeintervals(1,i) - 0.0 ) < tolerance)) .or.            &
+              (obs_hour ==  9 .and. (abs(timeintervals(1,i) - 0.25) < tolerance)) .or.            &
```

Comment:
Does timeintervals(1,:) refer to the **start** times of the observation interval?  
If so, I think we assimilate the obs from a "forward-looking" obs interval during a "backward-looking" assimilation window.  For example, if date_time=1:30z and the assimilation window is [0,3], then we set dt_obs=3 and assimilate obs taken during [3,9].  Can this be improved by changing the < and >= signs to <= and > in the calc of dt_obs?  I'm not really sure, but it's probably worth thinking about a bit more

Reply:
Yes, timeintervals(1,:) refers to the start times of the observation interval. 
I think the behaviour at the moment for 3hr assimilation windows centered on 1:30z and 4:30z is that it will assimilate all the obs from the first time slice (0 - 6hrs) in the first assimilation window centered on 1:30z and not in the second assimilation window centered on 4:30z. Changing < and >= signs to <= and > would reverse the behaviour.